server/language-redirect.js

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Included for types only.
+// eslint-disable-next-line no-unused-vars
+const express = require('express');
+
+const path = require('path');
+const fs = require('fs');
+
+const Negotiator = require('negotiator');
+
+const {rootDir, defaultLocale, doRedirect} = require('./env');
+
+const ignoredPathPrefixes = [
+  '_static',
+  'fonts',
+  'images',
+  'js',
+  'css',
+  'feeds',
+];
+
+/**
+ * @type {express.RequestHandler}
+ */
+const languageRedirectHandler = (req, res, next) => {
+  const pathParts = req.path.substring(1).split('/');
+  const pathPrefix = pathParts[0];
+
+  // If the request path is is a root level path, starts with two chars
```

Comment:
Maybe instead of heuristic based on number of letters etc you can use supported locales from https://github.com/GoogleChrome/developer.chrome.com/blob/main/site/_data/site.json#L8 ?

Reply:
There was a comment/issue/whatever somewhere that things in server should not rely on things from site in order to be able to only ship server and dist directories to App Engine at some point, to improve build times. That's why I did it that weird way 😅  but refactored it as suggested.